### PR TITLE
helm: Add support for mounting existing TLS secrets with optional root CA

### LINKED ISSUE #11103: Allow Use of Self-deployed TLS Key/Cert in Teleport Cluster Chart
### What

This [discussion post](https://github.com/gravitational/teleport/discussions/11101) describes the context for this, but to summarize - it would be nice to be able to specify a Kubernetes secret to be used for the Teleport [key and cert file](https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-cluster/templates/config.yaml#L102) so that cert-manager does not need to be installed as a dependency, and, as is needed in our case, to enable use of something like a [Cloudflare Origin Certificate](https://blog.cloudflare.com/cloudflare-ca-encryption-origin/) when terminating TLS in Teleport.

### How

I would say the chart should allow for more flexibility around how the certificates are supplied. The easiest way I can think to do this would be to add a value that lets you specify a Kubernetes Secret name containing the TLS cert and key. The other option here might be to enable overriding of the `https_keypairs` configuration block, which could then be pointed at a certificate mounted via the `extraVolumes`/`extraVolumeMounts` chart values.

### Why

Mentioned most of it in the first section, but this would enable use of an existing k8s secret to be used for TLS.

### Workaround
None that I'm aware of.

### Additional Info

I'm still rather new to Teleport, but the [docs](https://goteleport.com/docs/kubernetes-access/helm/guides/gcp/) don't seem to mention any other reasons why cert-manager/Let's Encrypt are requirements for the cluster other than for generating the initial cert/key PEM files. If I'm missing something here, please let me know, but this would definitely ease adoption for us if we didn't have to add the overhead of managing a cert-manager deployment in addition to the cluster and if we were able to use Cloudflare's proxied DNS features (DOS protection, WAF rules etc...) which are only possible through the use of an origin certificate.

Also, I'm happy to PR the changes myself, just wanted to get some feedback on what the preference would be for exposing this kind of option before I went ahead with that.

### PR DESCRIPTION
**Pull request by webvictim**
This PR adds a `tls.existingSecretName` value to the `teleport-cluster` Helm chart which makes it easier to provide a custom TLS keypair for Teleport to use when securing its web UI while still using other chart functionality.

It also provides a `tls.existingSecretNameCA` value to mount a root/intermediate CA or other bundle into the pod using the `SSL_CERT_FILE` environment variable so that Teleport can validate the full certificate chain itself and prevent `User Message: unable to verify HTTPS certificate chain` errors in the pod logs.

Docs counterpart is #11306 (separate PR due to some local workflow issues, fixed now)

Fixes #11103

### CODE REVIEW

File: examples/chart/teleport-kube-agent/templates/deployment.yaml [hunk 1]
```diff
@@ -110,7 +110,7 @@ spec:
         {{- end }}
         {{- if .Values.extraEnv }}
         env:
-          {{- toYaml .Values.extraEnv | nindent 10 }}
+          {{- toYaml .Values.extraEnv | nindent 8 }}
```
Comment (zmb3) [inline examples/chart/teleport-kube-agent/templates/deployment.yaml:113]: Was this an unrelated bug?

File: examples/chart/teleport-cluster/values.yaml [hunk 2]
```diff
@@ -164,6 +167,19 @@ highAvailability:
...
+  # Name of an existing secret to use which contains a TLS keypair. Will automatically set the https_keypairs section in teleport.yaml.
+  # Create the secret in the same namespace as Teleport using `kubectl create secret tls my-tls-secret --cert=/path/to/cert/file --key=/path/to/key/file`
+  # See https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets for more information.
+  existingSecretName: ""
+  # (optional) Name of an existing secret to use which contains a CA or trust bundle in x509 PEM format.
+  # Useful for building trust when using intermediate certificate authorities.
+  # This will automatically set the SSL_CERT_FILE environment variable to trust the CA.
+  # Create the secret with `kubectl create secret generic --from-file=ca.pem=/path/to/root-ca.pem
+  # The filename inside the secret is important - it _must_ be ca.pem
+  existingSecretNameCA: ""
```
Comment (zmb3) [inline examples/chart/teleport-cluster/values.yaml:182]: nit: should this be named `existingSecretCAName` instead?

**Review by zmb3 — APPROVED** [review]

File: examples/chart/teleport-kube-agent/templates/deployment.yaml (hunk 1 above)
Comment (webvictim) [reply examples/chart/teleport-kube-agent/templates/deployment.yaml:113]: A semi-related formatting issue. I added `SSL_CERT_FILE` to the environment as part of this PR and it turned out that we were over-indenting anything else passed via `extraEnv`.

File: examples/chart/teleport-cluster/values.yaml (hunk 2 above)
Comment (webvictim) [reply examples/chart/teleport-cluster/values.yaml:182]: If that seems clearer I can change it. I think it was `existingCASecretName` in my first iteration which does parse a little better.

**Comment by mattmelgard** [comment]
Ran into some issues when deploying/testing this out this evening, part of which may or may not just be something missing on my end. Specifically - 

It seems the image `quay.io/gravitational/teleport:10.0.0-dev` either doesn't seem to exist or I don't have permission to pull it since I was getting image pull errors when deploying the chart from your branch:
```
  Normal   BackOff      1s (x2 over 29s)   kubelet            Back-off pulling image "quay.io/gravitational/teleport:10.0.0-dev"
```
To get past this I overrode the value using `teleportVersionOverride` and set it to `9.0.0`, but then encountered this error when I set the existing secret value to use my Cloudflare Origin certificate which was deployed as a k8s TLS secret in the cluster:

```
k logs teleport-86bf88644f-d7kf7
ERROR: "unable to verify HTTPS certificate chain in : \x1b[31mERROR: \x1b[0mWARNING:"

  The proxy you are connecting to has presented a certificate signed by a
  unknown authority. This is most likely due to either being presented
  with a self-signed certificate or the certificate was truly signed by an
  authority not known to the client.

  If you know the certificate is self-signed and would like to ignore this
  error use the --insecure flag.

  If you have your own certificate authority that you would like to use to
  validate the certificate chain presented by the proxy, set the
  SSL_CERT_FILE and SSL_CERT_DIR environment variables respectively and try
  again.

  If you think something malicious may be occurring, contact your Teleport
  system administrator to resolve this issue.
```

For reference, here's the values I deployed that with:

```
chartMode: gcp
clusterName: teleport.mattusaur.com
authenticationType: github
teleportVersionOverride: "9.0.0"
gcp:
  projectId: matt-melgard-personal
  backendTable: teleport-helm-backend
  auditLogTable: teleport-helm-events
  sessionRecordingBucket: mattmelgard-teleport-helm-sessions
highAvailability:
  replicaCount: 2
tls:
  existingSecretName: cloudflare-origin-cert
annotations:
  service:
    external-dns.alpha.kubernetes.io/hostname: "teleport.mattusaur.com,*.teleport.mattusaur.com"
```

And here's the origin certificate I created in my personal Cloudflare tenant.

Screen Shot 2022-03-29 at 10 52 47 PM

This may just be a quirk of Cloudflare's [Origin Certificates](https://developers.cloudflare.com/ssl/origin-configuration/origin-ca/), but I'm still kind of digging into it to be sure. That or maybe I do need to set those env vars?

**Comment by webvictim** [comment]
@mattmelgard Thanks for testing. I'm unsure exactly what Cloudflare's origin certificates look like, but my guess is that their chain of trust is `Root CA -> Intermediate CA -> Origin cert` and the bundle you download only contains the Intermediate CA and Origin certs. In this situation the Root CA cert would need to be added using `existingSecretNameCA` so that the pods know to trust it.

I'm going to do a little testing with this myself to confirm and maybe update the instructions. I have a feeling that the cert parser which backs Teleport's `https_keypairs` isn't capable of reading a file containing a concatenated "Origin cert/Intermediate CA cert/Root CA cert" set, which is why we need this workaround. @zmb3 Any thoughts on this?

**Comment by webvictim** [comment]
@mattmelgard Looks like it works if you download Cloudflare's origin root CA (as per these instructions: https://developers.cloudflare.com/ssl/origin-configuration/origin-ca/#4-required-for-some-add-cloudflare-origin-ca-root-certificates - get one named "Cloudflare Origin RSA PEM") and add it as a CA secret:

```bash
wget https://developers.cloudflare.com/ssl/static/origin_ca_rsa_root.pem
kubectl create secret generic cloudflare-origin-cert-root-ca --from-file=ca.pem=origin_ca_rsa_root.pem
```

Add this to your values and do a `helm upgrade`:

```yaml
tls:
  existingSecretNameCA: cloudflare-origin-cert-root-ca
```

**Comment by mattmelgard** [comment]
> @mattmelgard Looks like it works if you download Cloudflare's origin root CA (as per these instructions: https://developers.cloudflare.com/ssl/origin-configuration/origin-ca/#4-required-for-some-add-cloudflare-origin-ca-root-certificates - get the one named "Cloudflare Origin RSA PEM") and add it as a CA secret:
> 
> ```shell
> wget https://developers.cloudflare.com/ssl/static/origin_ca_rsa_root.pem
> kubectl create secret generic cloudflare-origin-cert-root-ca --from-file=ca.pem=origin_ca_rsa_root.pem
> ```
> 
> Add this to your values and do a `helm upgrade`:
> 
> ```yaml
> tls:
>   existingSecretNameCA: cloudflare-origin-cert-root-ca
> ```
> 
> The issue with the `10.0.0-dev` image not existing is expected when deploying from a PR forked from `master` - i should have warned you. Using `teleportVersionOverride` was the correct approach.

Thanks for getting back so quickly, I appreciate the update!

Seems you have it working so that's awesome. I'll validate on my end as well later tonight just to give you the extra data point, but if all goes well then this seems like a perfect solution for us. Thanks again for digging into this!

**Comment by mattmelgard** [comment]
@webvictim were you able to get that solution to work with a Cloudflare proxied DNS record? Seems the pods stand up okay, but I'm getting a 522 error when Cloudflare tries to connect to my backend load balancer for some reason. Might just be something misconfigured on my end, but the setup I have doesn't _seem_ any different from other GCP managed endpoints that I've exposed using the Cloudflare origin certs/proxied DNS in the past. I'll try to dig in and debug a little more tomorrow though.

Screen Shot 2022-03-31 at 12 26 20 AM

**Comment by webvictim** [comment]
@mattmelgard Afraid I haven't been that far down the rabbit hole with Cloudflare - if the pod started once the root CA was provided then that pretty much satisfies my requirement. I suspect it might be some kind of configuration issue somewhere. Do let me know how you get on, and thanks for the existing testing!

File: examples/chart/teleport-cluster/values.yaml (hunk 2 above)
Comment (webvictim) [reply examples/chart/teleport-cluster/values.yaml]: Changed it to `existingCASecretName` which I think makes more sense.

**Review by russjones — APPROVED** [review]

**Pull request merged**